File: app/src/main/java/com/ghostnexora/vpn/tunnel/TlsTransport.java
package com.ghostnexora.vpn.tunnel;

import com.ghostnexora.vpn.data.model.TlsVerificationMode;

import java.net.Socket;
import java.util.Collections;

import javax.net.ssl.SNIHostName;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSocket;

/**
 * Single TLS construction path for VPN transports.
 *
 * The platform TrustManager remains authoritative. CUSTOM_SNI changes only the
 * hostname verification policy already represented by TlsVerificationMode; it
 * never installs a trust-all certificate manager.
 */
public final class TlsTransport {
    private static final String HTTPS_ENDPOINT_IDENTIFICATION = "HTTPS";

    private TlsTransport() {
    }

    public static SSLSocket upgrade(
            Socket connectedSocket,
            String targetHost,
            int targetPort,
            String sniHost,
            TlsVerificationMode verificationMode
    ) throws Exception {
        if (targetHost == null || targetHost.trim().isEmpty()) {
            throw new IllegalArgumentException("El host TLS no puede estar vacío");
        }
        if (targetPort < 1 || targetPort > 65_535) {
            throw new IllegalArgumentException("El puerto TLS es inválido");
        }
        if (sniHost == null || sniHost.trim().isEmpty()) {
            throw new IllegalArgumentException("El SNI no puede estar vacío");
        }
        if (connectedSocket == null) {
            throw new IllegalArgumentException("connectedSocket == null");
        }
        if (verificationMode == null) {
            throw new IllegalArgumentException("verificationMode == null");
        }

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, null, null);

        String verificationHost = verificationMode.getVerifiesHostname() ? sniHost : targetHost;
        SSLSocket sslSocket = (SSLSocket) sslContext.getSocketFactory().createSocket(
                connectedSocket,
                verificationHost,
                targetPort,
                true
        );
        sslSocket.setUseClientMode(true);
        sslSocket.setSSLParameters(configureParameters(
                sslSocket.getSSLParameters(),
                sniHost,
                verificationMode
        ));
        sslSocket.startHandshake();
        return sslSocket;
    }

    public static SSLParameters configureParameters(
            SSLParameters current,
            String sniHost,
            TlsVerificationMode verificationMode
    ) {
        if (current == null) {
            throw new IllegalArgumentException("current == null");
        }
        if (sniHost == null || sniHost.trim().isEmpty()) {
            throw new IllegalArgumentException("El SNI no puede estar vacío");
        }
        if (verificationMode == null) {
            throw new IllegalArgumentException("verificationMode == null");
        }

        current.setEndpointIdentificationAlgorithm(
                verificationMode.getVerifiesHostname() ? HTTPS_ENDPOINT_IDENTIFICATION : null
        );
        try {
            current.setServerNames(Collections.singletonList(new SNIHostName(sniHost)));
        } catch (Throwable error) {
            throw new IllegalArgumentException("SNI TLS inválido: " + sniHost, error);
        }
        return current;
    }
}

File: app/src/main/java/com/ghostnexora/vpn/tunnel/TlsTransport.kt
package com.ghostnexora.vpn.tunnel

import com.ghostnexora.vpn.data.model.TlsVerificationMode
import java.net.Socket
import javax.net.ssl.SNIHostName
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLParameters
import javax.net.ssl.SSLSocket

/**
 * Único punto de construcción TLS para SSH y diagnósticos.
 *
 * El socket siempre usa los TrustManager de la plataforma. La política
 * CUSTOM_SNI cambia únicamente la verificación de hostname para permitir que
 * el SNI y el SAN del certificado sean distintos.
 */
object TlsTransport {
    fun upgrade(
        connectedSocket: Socket,
        targetHost: String,
        targetPort: Int,
        sniHost: String,
        verificationMode: TlsVerificationMode
    ): SSLSocket {
        require(targetHost.isNotBlank()) { "El host TLS no puede estar vacío" }
        require(targetPort in 1..65535) { "El puerto TLS es inválido" }
        require(sniHost.isNotBlank()) { "El SNI no puede estar vacío" }

        val sslContext = SSLContext.getInstance("TLS")
        sslContext.init(null, null, null)

        val verificationHost = if (verificationMode.verifiesHostname) sniHost else targetHost
        val sslSocket = sslContext.socketFactory.createSocket(
            connectedSocket,
            verificationHost,
            targetPort,
            true
        ) as SSLSocket
        sslSocket.useClientMode = true
        sslSocket.sslParameters = configureParameters(
            current = sslSocket.sslParameters,
            sniHost = sniHost,
            verificationMode = verificationMode
        )
        sslSocket.startHandshake()
        return sslSocket
    }

    internal fun configureParameters(
        current: SSLParameters,
        sniHost: String,
        verificationMode: TlsVerificationMode
    ): SSLParameters = current.apply {
        endpointIdentificationAlgorithm =
            if (verificationMode.verifiesHostname) HTTPS_ENDPOINT_IDENTIFICATION else null
        serverNames = listOf(
            runCatching { SNIHostName(sniHost) }
                .getOrElse { throw IllegalArgumentException("SNI TLS inválido: $sniHost", it) }
        )
    }

    private const val HTTPS_ENDPOINT_IDENTIFICATION = "HTTPS"
}
